### ISSUE TITLE
[question] How to obtain package reference in Conan 2 during upload to Artifactory, similar to Conan 1.x ?

### ISSUE THREAD

**Issue by ATS-AValli** [conan-io/conan#14040]
### What is your question?

Hi.
In Conan 1.x, when an upload to Artifactory was done, it was possible to dump the package information with the option "--json" (documented here: [Upload command](https://docs.conan.io/1/reference/commands/creator/upload.html)).

Is it possible with Conan 2.x to obtain the same kind of information ? We're interested in the uploaded package reference used in Artifactory.

I've had a look at the [art:build-info](https://github.com/conan-io/conan-extensions/tree/main/extensions/commands/art) extension but I'm not sure it's the right kind of command.

Moreover, it shall be noted that for historical reasons (or hysterical raisins), the CI scheme into we're integrating Conan 2.x normally does a sequence of `conan install ... - conan build ... - conan upload`, thus no `conan create`
Any ideas ?

Thanks.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @ATS-AValli 

Thanks for your question. I like it because we just merged the ``pkg-lists`` feature for next Conan 2.0.7 that implements a better flow for this. So it will be:

```
conan create . .... --format=json > graph.json  # stores the created dependency graph in json
conan list --graph=graph.json --format=json > pkglist.json # creates a list of packages
conan upload --list=pkglist.json -r=myremote -c
```
The list of packages that is uploaded can be created including all dependencies, only the packages that were built from source, but not the dependencies, etc, there are some possibilities.

We will write a blog post about it, but if you want to see the PR https://github.com/conan-io/conan/pull/13928, or the docs in my branch in: https://github.com/conan-io/docs/compare/release/2.0...memsharded:docs:feature/pkglists

**Comment by ATS-AValli** [comment]
Hi @memsharded 
Thanks a lot, we'll be waiting for 2.0.7 .

Just a question: we use the sequence `conan install` + `conan build` instead of `conan create` as your example above. Would it work as well ? I suppose so, but just to be sure.

**Comment by memsharded** [comment]
Yes ``conan install`` or ``conan build`` (if you are doing the second, you don't need the first, it is included) will also be able to create a ``graph.json`` file that can be used to create a package list to be uploaded. Just take into account that the thing that you are building is just local, so it is not a Conan package and will not be uploaded, we would be talking about the dependencies of that thing only. Does it make sense?

**Comment by ATS-AValli** [comment]
@memsharded 

> [...] Just take into account that the thing that you are building is just local, so it is not a Conan package and will not be uploaded, we would be talking about the dependencies of that thing only. Does it make sense?

Hmmm... I'm not sure this is what we are looking for then. We need the same kind of information, namely the package reference in Artifactory, that was available with conan 1.x using the command `conan upload --json=filename` .
It's what the 1.x documentation refers as: `json file path where the upload information will be written to` .

This is what we need. Is there a way to have the same information on V2.x ?

Thanks!

**Comment by memsharded** [comment]
But isn't the reference uploaded to artifactory exactly the same as the reference you are uploading?, it should be the same.

I thought that you wanted to know the packages that "need to be uploaded" after a create/install, but it doesn't seem the case. We can definitely add the json output for the ``upload`` command if that is what you are looking for, but it wouldn't be different to the one doing a ``conan list --format=json>pkglist.json`` and then feeding that json to ``conan upload --list=pkglist.json``. Maybe I am missing to understand something?

**Comment by ATS-AValli** [comment]
> But isn't the reference uploaded to artifactory exactly the same as the reference you are uploading?, it should be the same.

I was told by the person who manages our CI that is something he gets from the JSON of Conan 1.x with conan upload. It's not the package reference but some form of hash by which Artifactory identifies the package.
I've asked him to provide an example and I'll ask here again once I have the example.

> I thought that you wanted to know the packages that "need to be uploaded" after a create/install, but it doesn't seem the case. We can definitely add the json output for the `upload` command if that is what you are looking for, but it wouldn't be different to the one doing a `conan list --format=json>pkglist.json` and then feeding that json to `conan upload --list=pkglist.json`. Maybe I am missing to understand something?

It's not about the packages that were used to be built the current package, but the Artifactory "key" for the newly uploaded package.

**Comment by ATS-AValli** [comment]
@memsharded 
Update: from what I've been told, it's what in the new SQLite DB is in the columns `rrev`, `pkgid` and `prev` , all three of them.
In the old JSON output for upload, it's what was in the path `uploaded - packages - id` .

Is there a standard way to have this information ?

**Comment by memsharded** [comment]
> I was told by the person who manages our CI that is something he gets from the JSON of Conan 1.x with conan upload. It's not the package reference but some form of hash by which Artifactory identifies the package.

I think there is some misunderstanding. Artifactory (or any other server) do not assign any new identifier or hash. Everything (recipe-revision, package-id, package-revision) is already defined in the client side after a ``conan install --build`` or ``conan create``. 

> It's not about the packages that were used to be built the current package, but the Artifactory "key" for the newly uploaded package.

As commented above, that is not a thing, Artifactory does not assign any key to packages

> Update: from what I've been told, it's what in the new SQLite DB is in the columns rrev, pkgid and prev , all three of them.
In the old JSON output for upload, it's what was in the path uploaded - packages - id .

> Is there a standard way to have this information ?

Yes, definitely, all of it is in the ``json`` output:

```bash
conan create . --format=json

"1": {
                "ref": "pkg/1.0#b4c41cddd96b5279e7c5d422a2a4e53d",
                "id": "1",
                "recipe": "Cache",
                "package_id": "da39a3ee5e6b4b0d3255bfef95601890afd80709",
                "prev": "0ba8627bd47edc3a501e8f0eb9a79e5e",
                "build_id": null,
                "binary": "Build"
```

In general, you should be looking for things with ``binary: "Build"``, as those were built from source, you want to upload them, to avoid re-building

**Comment by ATS-AValli** [comment]
Ok, thanks very much for the clarification.
If we don't use `conan create` but only `conan install + conan build`, is the same information available ?

**Comment by memsharded** [comment]
``conan install`` + ``conan build`` does not create a package at all, so the thing that you build does not exist for conan, is just something in your local folder. But:

- In Conan 2.0, the extra ``conan install`` is not necessary, ``conan build`` already calls the install internally
- Both ``install`` and ``build`` commands, they have the ``--format=json`` output, with the details of the graph, what packages have been built from sources. So if you want to upload those packages to the server, yes, the same commands can be used.

**Comment by ATS-AValli** [comment]
Yes, sorry, I'm checking with the other person who's responsible for the CI: they do install + build + export_pkg I believe. Let me check...

**Comment by memsharded** [comment]
The ``export-pkg`` is similar to a ``create`` process, so it also has its ``--format=json`` output, that will contain the information about the ``package_id``, revisions, etc, necessary to identify the things to upload.

**Comment by ATS-AValli** [comment]
Hi.
So, the request I got from the person in charge of our CI is to have once again the output in JSON format during upload. I'm aware it doesn't make much sense to be so fixed on having it in the upload command rather than in export-pkg like you kindly suggest.

So, if you could also have it in upload, all the better. If not, just leave everything like it is, it's not a big issue.

Thanks for being so available.

**Comment by memsharded** [comment]
Good news :) 

The ``conan upload`` command also got ``--format=json`` output in https://github.com/conan-io/conan/pull/14002, merged today for next Conan 2.0.7. The output will be a json with the "pkg-list" format (same as ``conan list`` command)

**Comment by ATS-AValli** [comment]
Thanks a lot, that's above and beyond the call of duty. :-)
